### PR TITLE
KITE-1101: Added logic to FileSystemPartitionIterator to exclude dire…

### PR DESCRIPTION
**Pull request by mkwhitacre**
…ctories with data that does not qualify. Extended predicates and StorageKey code to support partial values.

Note this might violate some of the initial designs for StorageKey and the Predicates so I'm open to alternate ways of doing this logic.  @rdblue and @rbrush if you have some insight or critique that would be helpful.  It is not necessarily the cleanest changes so if you have suggestions on how to clean this up with original design intent that would be helpful.

In some smaller tests I've seen significant improvements in job startup time (1s vs 100s) and a lot less memory pressure on the driver program when kicking off the job.

### CODE REVIEW

File: kite-data/kite-data-core/src/main/java/org/kitesdk/data/spi/StorageKey.java [hunk 1]
```diff
@@ -132,8 +135,6 @@ public void replace(int index, Object value) {
    */
   public void replaceValues(List<Object> values) {
     Preconditions.checkArgument(values != null, "Values cannot be null");
-    Preconditions.checkArgument(values.size() == fields.size(),
```
Comment (noslowerdna) [inline kite-data/kite-data-core/src/main/java/org/kitesdk/data/spi/StorageKey.java]: Would it make sense to change this to <= as a sanity check?

File: kite-data/kite-data-core/src/main/java/org/kitesdk/data/spi/filesystem/FileSystemPartitionIterator.java [hunk 2]
```diff
@@ -62,15 +66,28 @@ public FileSystemIterator(int depth) throws IOException {
...
+      if(key.size() > 0){
+        applies = predicate.apply(key);
+      }else{
+        LOG.debug("Not applying predicate proactively because path {} does not have any key values.", dir);
+      }
+      if(applies) {
+        try {
+          for (FileStatus stat : fs.listStatus(dir, PathFilters.notHidden())) {
+            if (stat.isDir()) {
+              // TODO: add a check here for range.couldContain(Marker)
```
Comment (noslowerdna) [inline kite-data/kite-data-core/src/main/java/org/kitesdk/data/spi/filesystem/FileSystemPartitionIterator.java]: Is this TODO comment still applicable?

File: kite-data/kite-data-core/src/main/java/org/kitesdk/data/spi/filesystem/FileSystemViewKeyInputFormat.java [hunk 3]
```diff
@@ -19,6 +19,7 @@
 import com.google.common.collect.Lists;
 import java.io.IOException;
 import java.util.Iterator;
+import java.util.LinkedList;
```
Comment (noslowerdna) [inline kite-data/kite-data-core/src/main/java/org/kitesdk/data/spi/filesystem/FileSystemViewKeyInputFormat.java]: Looks like you can revert the changes in this file assuming all unintentional.

File: kite-data/kite-data-core/src/main/java/org/kitesdk/data/spi/filesystem/FileSystemPartitionIterator.java (hunk 2 above)
Comment (mkwhitacre) [inline kite-data/kite-data-core/src/main/java/org/kitesdk/data/spi/filesystem/FileSystemPartitionIterator.java]: That's a good question.  That TODO was there prior to the changes and I'm not sure I completely grokked the optimization.

File: kite-data/kite-data-core/src/main/java/org/kitesdk/data/spi/filesystem/FileSystemViewKeyInputFormat.java (hunk 3 above)
Comment (mkwhitacre) [inline kite-data/kite-data-core/src/main/java/org/kitesdk/data/spi/filesystem/FileSystemViewKeyInputFormat.java]: Yep, I put some timings in this class and didn't get them cleanly removed.  I'll revert this class changes.

File: kite-data/kite-data-core/src/main/java/org/kitesdk/data/spi/StorageKey.java (hunk 1 above)
Comment (mkwhitacre) [inline kite-data/kite-data-core/src/main/java/org/kitesdk/data/spi/StorageKey.java]: Follow up commit should have this sanity check.

File: kite-data/kite-data-core/src/main/java/org/kitesdk/data/spi/filesystem/FileSystemViewKeyInputFormat.java (hunk 3 above)
Comment (mkwhitacre) [inline kite-data/kite-data-core/src/main/java/org/kitesdk/data/spi/filesystem/FileSystemViewKeyInputFormat.java]: Rolled back this file.

**Comment by noslowerdna** [comment]
+1

**Comment by rbrush** [comment]
+1. Looks good to me.

**Comment by mkwhitacre** [comment]
Got two +1s.  Merging.
